File: backend/app/domain/lane_pressure.py
"""Lane pressure domain models for map control tracking."""

from typing import Optional
from sqlmodel import SQLModel


class LanePressureSnapshot(SQLModel):
    """Snapshot of lane pressure at a given second.

    Pressure represents how far into enemy territory the wave has pushed,
    scaled by alive creep count.
    - 0.0 = wave dead or at own base
    - >0.0 = wave alive; higher values mean closer to enemy objective with more creeps
    """

    pressure: float               # raw_pressure * (alive_count * 0.25)
    team: int                     # Team (2 = Amber, 3 = Sapphire)
    wave_id: str                  # "lane_team_spawnsec"
    creep_count: int              # alive creep count at this second
    attributed_players: list[int] # union of nearby_players across alive creeps


class LanePressureData(SQLModel):
    """Container for all lane pressure data.

    Key: wave_id (same keys as LaneCreepData.wave_meta)
    Value: per-second snapshots (None = no alive creeps that second)
    """

    pressure: dict[str, list[Optional[LanePressureSnapshot]]]

File: frontend/src/domain/lanePressure.ts
// Lane pressure types for map control tracking

export interface LanePressureSnapshot {
  pressure: number; // 0-1 value (0 = own base, 1 = enemy base)
  team: number; // Team (2 = Amber, 3 = Sapphire)
  wave_id: string;
  creep_count: number;
  attributed_players: number[]; // Player custom_ids within proximity
}

// Key format: wave_id (e.g., "1_2_45" for lane 1, team 2, spawned at sec 45)
// Value: per-second snapshots (null if wave not active that second)
export interface LanePressureData {
  pressure: Record<string, (LanePressureSnapshot | null)[]>;
}
